graphql_compiler/cost_estimation/cardinality_estimator.py

```diff
@@ -69,17 +69,103 @@ def _get_last_edge_direction_and_name_to_location(location):
+
+
+def _get_outbound_and_inbound_names_from_locations(query_metadata, child_location, parent_location):
+    """Get the out and in classnames of a location and its parent from the last edge information."""
+    edge_direction, edge_name = _get_last_edge_direction_and_name_to_location(child_location)
+    parent_name_from_location = query_metadata.get_location_info(parent_location).type.name
+    child_name_from_location = query_metadata.get_location_info(child_location).type.name
+    if edge_direction == INBOUND_EDGE_DIRECTION:
+        outbound_vertex_name = child_name_from_location
+        inbound_vertex_name = parent_name_from_location
+    elif edge_direction == OUTBOUND_EDGE_DIRECTION:
+        outbound_vertex_name = parent_name_from_location
+        inbound_vertex_name = child_name_from_location
+    else:
+        raise AssertionError(u'Expected edge direction to be either inbound or outbound.'
+                             u'Found: edge {} with direction {}'.format(edge_name, edge_direction))
+    return outbound_vertex_name, inbound_vertex_name
+
+
+def _probe_statistics_for_vertex_edge_vertex_count(
+    statistics, query_metadata, child_location, parent_location
+):
+    """Probe for the count of edges that connect parent_location and child_location vertices."""
+    _, edge_name = _get_last_edge_direction_and_name_to_location(child_location)
+    outbound_vertex_name, inbound_vertex_name = _get_outbound_and_inbound_names_from_locations(
+        query_metadata, child_location, parent_location
+    )
+    probe_result = statistics.get_vertex_edge_vertex_count(
+        outbound_vertex_name, edge_name, inbound_vertex_name
+    )
+    return probe_result
+
+
+def _estimate_vertex_edge_vertex_count_using_class_count(
+    schema_graph, statistics, query_metadata, child_location, parent_location
+):
+    """Estimate the count of edges that connect parent_location and child_location vertices.
+
+    Given a parent location of class A and a child location of class B, this function estimates the
+    number of AB edges using class counts. If A and B are subclasses of the edge's endpoint classes
+    (which we'll name C and D respectively), we only have statistics for CD edges. So estimates for
+    the number of AB edges will be made using the assumption that CD edges are distributed
+    independently of whether or not the vertex of class C is also of class A and likewise for D and
+    B. In the general case, we estimate the statistic as
+    (# of AB edges) = (# of CD edges) * (# of A vertices) / (# of C vertices) *
+                                        (# of B vertices) / (# of D vertices).
+
+    Args:
+        schema_graph: SchemaGraph object
+        statistics: Statistics object
+        query_metadata: QueryMetadataTable object
+        child_location: BaseLocation object
+        parent_location: BaseLocation object
+
+    Returns:
+        float, estimate for (# of AB edges)
```

Comment:
Please use precise language rather than relying on symbols, unless you've defined the symbol previously in the docstring. Replace `#` with `number` etc.